### ISSUE TITLE
Pretty-printing of decoding failure history

### ISSUE THREAD

**Issue by ArtemPyanykh** [travisbrown/circe#188]
Consider the following example:

```scala
scala> val json = parse("""[1, 2, "hello", 4, "world", 6]""").getOrElse(Json.empty)
scala> Decoder[List[Int]].apply(json.hcursor)
res5: io.circe.Decoder.Result[List[Int]] = Left(io.circe.DecodingFailure: Int: El(MoveRight,true),El(MoveRight,true),El(DownArray,true))
```

This part might be hard to parse for a human eye:
```scala
io.circe.DecodingFailure: Int: El(MoveRight,true),El(MoveRight,true),El(DownArray,true)
```

I guess we could improve the error message by reversing a history, making it a bit more visual, and maybe adding the element we failed to parse. Something along these lines:
```
[↴ array]
  →|→ × Failed to decode hello as Int
```

I'm not sure about arrow symbols though (on github they are very tiny), so maybe just printing
```
[down array]
  [right], [right] × Failed to decode hello as Int
```

**Comment by ArtemPyanykh** [comment]
I can take a look into it in a couple weeks after I get back from vacation.

**Comment by travisbrown** [comment]
:+1:, this sounds great.

**Comment by tobias-johansson** [comment]
**WIP**
I've done some work on a `Show[DecodingFailure]` that tries to make simple, compact messages by representing the cursor history as javascript style selection, i.e. `"DecodingFailure at .foo.bar[2].baz: some message"`.
I believe this is a good idea if you want to return error messages to http clients, etc.

WIP can be seen here: https://github.com/travisbrown/circe/compare/master...tobias-johansson:Show-DecodingFailure
It deals with normal navigation operations, and if there are other kinds of operations it falls back to `Show[CursorOp]` for those.

**Problem**
The error messages could still be improved though, and I would like to discuss if it's possible to encode the reason for failure in `DecodingFailure`s in a better way.
At the moment if I do
```scala
scala> decode[Int](""""a"""")
res8: cats.data.Xor[io.circe.Error,Int] = Left(io.circe.DecodingFailure: Int)
```
the message in the `DecodingFailure` is simply the string `"Int"` which isn't of much help when trying to create a nice error message.

**Suggestion**
I'm suggesting an explicit encoding of the reason to replace or complement the `message` field in `DecodingFailure`, something like:
```scala
sealed trait Reason
case class TypeExpectation(expected: String) extends Reason
case class FailedValidation(message: String) extends Reason
case object FailedCursor extends Reason
```

This would make it easy to create or customize error messages, or otherwise interpret failures, without the need to embed the full messages in `Decoder` etc.
The downside might be that `DecodingFailure` is public api along with some other things like `Decoder[A].withErrorMessage` and I'm not sure how to deal with that. 

What do you think?

**Comment by ArtemPyanykh** [comment]
@tobias-johansson nice! I had similar idea in regard to implementation, but the JS-selection-like output scales better.

**Comment by travisbrown** [comment]
Quoting myself from Gitter:

> @tobias-johansson That sounds like reasonable output for `Show[DecodingFailure]`. I think it'd make sense to have the visualization that @ArtemPyanykh describes in #188 be available as something like a draw method on `DecodingFailure`, while the `Show` instance gives the more compact representation you describe.

So while #222 is merged, I think we should leave this issue open for the original more visual output proposed by @ArtemPyanykh.

**Comment by ArtemPyanykh** [comment]
I see that `CursorOp` already has [`Show` instance](https://github.com/travisbrown/circe/blob/master/core/shared/src/main/scala/io/circe/CursorOp.scala) defined, and string representations of operators are already pretty visual there. 

@tobias-johansson in his PR:
1. collapsed array traversals which is very reasonable since you hardly ever want to see 27 arrows on the screen `->` instead of `[27]`, and 
2. changed the repr. of `downField` from `--\\(field)"` to `.field`.

These 2 things probably covers 80% of use-cases. For other Ops it just fallbacks to the already defined Show. Which gives us pretty much what I wanted to do minus a couple extra line-breaks :smiley:

@travisbrown do you think that going for the original output format is still relevant (comparing them side by side)?

```scala
[↴ array]
  →|→|→|→|→|→|→|→|→|→|→|→ × Int
// and
Decoding Failure at [12]: Int
```

From https://github.com/travisbrown/circe/pull/222#issuecomment-194017772,
```scala
DecodingFailure at .bar.baz.qux.a: String
``` 
looks really nice! What might be cool though is adding a current focus to the error message so it gives a bit more context
```scala
DecodingFailure at .bar.baz.qux.a: decoding 10 as String
```
